### PR TITLE
Same54x atecc608 support

### PR DESCRIPTION
**Pull request by nik-markovic**
Add support for a new Microchip Azure RTOS version for SAM E54 Xplained Pro with ATECC608 support.

### CODE REVIEW

File: iotc-azrtos-sdk/authentication/driver/to_auth_driver.c [hunk 1]
```diff
@@ -99,18 +99,14 @@ static NX_CRYPTO_KEEP UINT  to_ecdsa_operation(UINT op,      /* Encrypt, Decrypt
 		static uint8_t s_r_buffer[SR_BUFFER_SIZE];
 		TO_ret_t ret = TOSE_sign_hash(to_context->to_ctx, to_context->driver_parameters.operational_identity_slot, input, s_r_buffer);
 		if (TORSP_SUCCESS != ret) {
-			printf("ADTO: TOSE_sign error 0x%04X\r\n", ret);
+	    	printf("ADTO: TOSE_sign error 0x%04X\r\n", ret);
```
Comment (nbeck-witekio) [inline iotc-azrtos-sdk/authentication/driver/to_auth_driver.c]: Indentatation

File: iotc-azrtos-sdk/authentication/driver/to_auth_driver.c [hunk 2]
```diff
@@ -99,18 +99,14 @@ static NX_CRYPTO_KEEP UINT  to_ecdsa_operation(UINT op,      /* Encrypt, Decrypt
...
 			return NX_CRYPTO_NOT_SUCCESSFUL;
 		}
-
-		// Uncommenting this section may help debugging but is a security risk:
-#if 0
 		printf("ADTO: ECDSA(");
```
Comment (nbeck-witekio) [inline iotc-azrtos-sdk/authentication/driver/to_auth_driver.c:105]: Should this be printed out?

File: iotc-azrtos-sdk/src/iotconnect_di.c [hunk 3]
```diff
@@ -23,7 +21,9 @@
 
 #define URL_RESOUCE_BUFFER_SIZE 40
 
-#define WORK_BUFFER_SIZE (TO_CERT_X509_MAXSIZE * 2 + 1)
+// some certs can be more than 512 bytes, so we ought to have enough space for all of 
+// it the hex characters. 
+#define WORK_BUFFER_SIZE (1500)
```
Comment (nbeck-witekio) [inline iotc-azrtos-sdk/src/iotconnect_di.c]: Would it be better to extend TO_CERT_X509_MAXSIZE to an appropriate larger size (not sure if something else depends on it being a specific size though?), or create a new larger define

File: samples/same54xprov2/basic-sample/include/app_config.h [hunk 4]
```diff
@@ -0,0 +1,25 @@
...
+// Copyright: Avnet 2021
+// Created by Nik Markovic <nikola.markovic@avnet.com> on 4/19/21.
+//
+
+#define IOTCONNECT_APP_USE_WEATHER_CLICK
+
+#ifndef APP_CONFIG_H
+#define APP_CONFIG_H
+
+#define IOTCONNECT_CPID "avtds"
```
Comment (nbeck-witekio) [inline samples/same54xprov2/basic-sample/include/app_config.h]: Remove this value and squash commits

File: samples/same54xprov2/basic-sample/include/app_config.h [hunk 5]
```diff
@@ -0,0 +1,25 @@
...
+#define IOTCONNECT_ENV  "avnetpoc"
```
Comment (nbeck-witekio) [inline samples/same54xprov2/basic-sample/include/app_config.h]: Remove this value and squash commits

File: samples/same54xprov2/basic-sample/include/app_config.h [hunk 6]
```diff
@@ -0,0 +1,25 @@
...
+
+// If using the ENABLE_DDIM_PKCS11_ATCA_DRIVER_SAMPLE define, the DUID will be automatically generated
+//#define IOTCONNECT_DUID "same54xpro" // you can supply a custom device UID, or...
```
Comment (nbeck-witekio) [inline samples/same54xprov2/basic-sample/include/app_config.h]: Would using a specific device id be easier for a demo? IMHO it doesn't make much difference if it is  IOTCONNECT_DUID or DUID_PREFIX as long as it is documented,  but DUID_PREFIX has the advantage of multiple devices not needing a new binary compiled  :)

File: samples/same54xprov2/basic-sample/src/iotconnect_app.c [hunk 7]
```diff
@@ -0,0 +1,310 @@
...
+extern UCHAR _nx_driver_hardware_address[];
+static IotConnectAzrtosConfig azrtos_config;
+static IotcAuthInterfaceContext auth_driver_context = NULL;
+
+#ifdef ENABLE_DDIM_PKCS11_ATCA_DRIVER_SAMPLE
+static char duid_buffer[IOTC_COMMON_NAME_MAX_LEN]; // from ATECC608 common name
+#endif
+#define APP_VERSION "01.00.00"
+
+//#define MEMORY_TEST
```
Comment (nbeck-witekio) [inline samples/same54xprov2/basic-sample/src/iotconnect_app.c:30]: Remove this and pass it in via -DMEMORY_TEST if wanting to build the memory test would be a tiny bit tidier

File: samples/same54xprov2/basic-sample/src/iotconnect_app.c [hunk 8]
```diff
@@ -0,0 +1,310 @@
...
+		   	auth_context, //
+			&cert, //
+			&cert_size //
+			);
+    if (0 == cert_size) {
+        printf("Unable to get the certificate from the driver.\r\n");
+        pkcs11_atca_release_auth_driver(auth_context);
+        return false;
+    }
+    char* b64_string_buffer = malloc(1024); // cert should be around 500, so this should not exceed 1024
```
Comment (nbeck-witekio) [inline samples/same54xprov2/basic-sample/src/iotconnect_app.c]: Magic number

File: samples/same54xprov2/basic-sample/src/iotconnect_app.c [hunk 9]
```diff
@@ -0,0 +1,310 @@
...
+    if (NULL == b64_string_buffer) {
+        pkcs11_atca_release_auth_driver(auth_context);
+        printf("WARNING: Unable to allocate memory to display the certificate.\r\n");
+    } else {
+        size_t b64_string_len = 1024; // in/out parameter. See atcacert_encode_pem_cert()
```
Comment (nbeck-witekio) [inline samples/same54xprov2/basic-sample/src/iotconnect_app.c]: Use same define as recommended for line239

File: samples/same54xprov2/basic-sample/src/iotconnect_app.c (hunk 8 above)
Comment (nbeck-witekio) [reply samples/same54xprov2/basic-sample/src/iotconnect_app.c]: If this is just a base64 encoded version or ASCII armored PEM version of the certificate then we can calculate the size at compile-time based on the size of the variable for holding the certificate

File: samples/same54xprov2/basic-sample/src/iotconnect_app.c [hunk 10]
```diff
@@ -0,0 +1,310 @@
...
+    }
+
+    printf("Obtained device certificate:\r\n\r\n");
+    char* operational_cn = ddim_interface.extract_operational_cn(auth_context);
+    if (NULL == operational_cn) {
+        pkcs11_atca_release_auth_driver(auth_context);
+        printf("Unable to get the certificate common name.\r\n");
+        return false;
+    }
+    strcpy(duid_buffer, operational_cn);
```
Comment (nbeck-witekio) [inline samples/same54xprov2/basic-sample/src/iotconnect_app.c:260]: Use a safe strcpy variant e.g. strcpy_s if available or strncpy with checks

**Review by nbeck-witekio — COMMENTED** [review]
Use safe/secure variants of string manipulation functions
Use secure version of memset (one that ensures it is not optimised away). 
There are a lot of commit messages that are not particularly descriptive that could be squashed/rebased before creating the PR, but if the intent is to squash commits when merged then this point is moot.
Magic numbers are used for a number of return values, it would be good to use a set of error/return codes, maybe either threadx or define a set of iotconnect return codes.
There are some good comments in the code, it would be leverage something like doxygen to generate API documentation.
A set of logging macros would provide some benefits over using printf directly everywhere. That way they could be easily enabled/disabled per module and different log levels could be defined to display or prevent displaying particular info without having to modify the code. These could use     __func__ etc instead of it being hardcoded in each of the prints

**Comment by nik-markovic** [comment]
> Use safe/secure variants of string manipulation functions
> Use secure version of memset (one that ensures it is not optimised away).

We will not be going that far in this project to convert all standard string/memset functions to their "secure" counterparts. Where necessary, buffer sizes and checks are performed manually and the user is informed. If there are specific places where this is not the case, overflows can predictably occur and overflows can be prevented, you can pinpoint those specific cases and we can address them on case-by-case basis. Same for memset. I do not understand the part about memset being "optimised away" though. Is that some potential exploit?

> There are a lot of commit messages that are not particularly descriptive that could be squashed/rebased before creating the PR, but if the intent is to squash commits when merged then this point is moot.

Yes, I usually squash the merge. I hope I don't mess up :) As I pointed out, this branch for some reason  I didn't do some merge correctly so too many commits seem to be a part of this PR.

> Magic numbers are used for a number of return values, it would be good to use a set of error/return codes, maybe either threadx or define a set of iotconnect return codes.

This would be way too much effort to do, but at some point if priorities allow and if we get a customer to warrant this effort, we should probably do it.  We return errors in so many cases that the error code list would be huge. However, we ALWAYS ensure that we or an underlying functions prints the error, so the customer knows exactly what the problem is without needing to print the error code and search for it in the code. The specific return codes are there (and hardcoded unfortunately) just so that the user can in **rare** cases do something specific when a specific error code is returned.

> There are some good comments in the code, it would be leverage something like doxygen to generate API documentation.

I am strictly against doxygen. It promotes garbage comments just for the sake of filling format. I'm sure you have seen in other projects where an obvious function like get_current_time(time_t* time) is "overcommented" with obvious comments like "returns the current time" and "@param time The time". Instead, we should se meaningful function and variable names, put in comments only for the cases where something is not already obvious/clear in the function description. This makes it possible to view 100 LOC of code on one screen with useful comments vs. having a 1000 LOC file filled with duxygen headers. You can see this philosophy in the contributing guidelines which should actually apply to all C SDK projects at https://github.com/avnet-iotconnect/iotc-c-lib/blob/master/CONTRIBUTING.md.

> A set of logging macros would provide some benefits over using printf directly everywhere. That way they could be easily enabled/disabled per module and different log levels could be defined to display or prevent displaying particular info without having to modify the code. These could use **func** etc instead of it being hardcoded in each of the prints

I definitely agree that SDK (and library code) should not use printf directly. This will also make it possble to remove the strings from code to reduce the SDK size in some cases. If I were to do this, then your first comment should apply (we need to have better error codes). I just feel that my time is better spent adding new content over something that's not an issue (yet) or "can" be improved. On a related note, reliance on time() function in the c iotc-c-lib and no ability to specify a custom malloc()/free() has been something that I wanted to address with higher priority.

If you feel strongly that some of these comments should still be implemented and want to ensure that it followed through, you can submit an issue and we can track/address it separately. It's not like we don't know that most of these comments (to which I agree with) are things that we can improve on, it's just that I didn't get around to taking care of it, or committing to supporting the infrastructure changes in the long term.

**Review by krantik — APPROVED** [review]

**Pull request merged**